File: algorithms/2_medium/T/01471_the_k_strongest_values_in_an_array/sort.cpp
typedef ptrdiff_t isize;

class Solution {
public:
    static vector<int> getStrongest(vector<int>& arr, int k) {
        assert(k >= 1 && k <= std::ssize(arr));
        const isize arr_len = std::ssize(arr);
        ranges::sort(arr);
        const isize mid = (arr_len - 1) / 2;
        const int m = arr[mid];
        auto res = vector<int>();
        res.reserve(k);

        isize lo = 0;
        isize hi = arr_len - 1;

        for (int step = 0; step < k; step += 1) {
            if (abs(arr[lo] - m) > abs(arr[hi] - m)) {
                res.push_back(arr[lo]);
                lo += 1;
            } else {
                res.push_back(arr[hi]);
                hi -= 1;
            }
        }

        return res;
    }
};

File: algorithms/2_medium/T/01471_the_k_strongest_values_in_an_array/sort.rs
impl Solution {
    pub fn get_strongest(mut arr: Vec<i32>, k: i32) -> Vec<i32> {
        assert!(k >= 1 && k as usize <= arr.len());
        if (k as usize == arr.len()) { return arr; }
        arr.sort_unstable();
        let mid = (arr.len() - 1) / 2;
        let m = arr[mid];
        let mut res: Vec<i32> = Vec::with_capacity(k as usize);

        let mut lo = 0_usize;
        let mut hi = arr.len() - 1;

        for _ in 0..k {
            if (arr[lo] - m).abs() > (arr[hi] - m).abs() {
                res.push(arr[lo]);
                lo += 1;
            } else {
                res.push(arr[hi]);
                hi -= 1;
            }
        }

        res
    }
}
